src/Learning/KWData/KWClass.cpp

```diff
@@ -2676,59 +2685,111 @@ boolean KWClass::CheckClassComposition(KWAttribute* parentAttribute, NumericKeyD
 
-			// La cle de la classe utilisee doit etre strictement plus longue que celle de la classe utilisante
-			// dans le cas d'un lien de composition multiple a trois niveau (ou plus)
-			// Exception dans le cas d'une classe sans-cle pouvant etre issue d'une regle de creation de table,
-			// auquel cas les cles ne sont pas necessaire y compris dans le cas d'un flocon creer par des regles
-			assert(parentClass == NULL or parentClass->GetKeyAttributeNumber() <= GetKeyAttributeNumber());
-			if (parentClass != NULL and parentAttribute->GetType() == KWType::ObjectArray and
-			    parentClass->GetKeyAttributeNumber() > 0 and
-			    parentClass->GetKeyAttributeNumber() >= GetKeyAttributeNumber())
+			// Verification des cle si demande
+			if (bOk and bCheckKeys)
 			{
-				AddError(sTmp + "As dictionary " + parentClass->GetName() + " owns variable " +
-					 parentAttribute->GetName() + " of type " +
-					 KWType::ToString(parentAttribute->GetType()) + "(" +
-					 parentAttribute->GetClass()->GetName() + ") and dictionary " +
-					 parentAttribute->GetClass()->GetName() + " itself owns variable " +
-					 attribute->GetName() + " of type " + KWType::ToString(attribute->GetType()) +
-					 "(" + attribute->GetClass()->GetName() + "), the key length in dictionary " +
-					 GetName() + " (key length = " + IntToString(GetKeyAttributeNumber()) +
-					 ") must be strictly greater than that of its parent "
-					 "dictionary " +
-					 parentClass->GetName() +
-					 " (key length = " + IntToString(parentClass->GetKeyAttributeNumber()) + ")");
-				bOk = false;
+				// La classe utilisee doit avoir une cle
+				if (attributeClass->GetKeyAttributeNumber() == 0)
+				{
+					if (bVerboseCheckKeys)
+						AddError(sTmp + "The dictionary related to variable " +
+							 attribute->GetName() + " of type " +
+							 RelationTypeToString(attribute) + " must have a key");
+					bOk = false;
+				}
+				// La cle de la classe utilisee doit etre au moins aussi longue que
+				// celle de la classe utilisante dans le cas d'un lien de composition
+				else if (attributeClass->GetKeyAttributeNumber() < GetKeyAttributeNumber())
+				{
+					if (bVerboseCheckKeys)
+						AddError(sTmp + "The key length (" +
+							 IntToString(attributeClass->GetKeyAttributeNumber()) +
+							 ") of the dictionary related to variable " +
+							 attribute->GetName() + " of type " +
+							 RelationTypeToString(attribute) +
+							 " must not be less than that of its parent dictionary " +
+							 GetName() + " (" + IntToString(GetKeyAttributeNumber()) + ")");
+					bOk = false;
+				}
+				// La cle de la classe utilisee doit etre strictement plus longue que celle de la classe utilisante
+				// dans le cas d'un lien de composition multiple a trois niveau (ou plus)
+				// Exception dans le cas d'une classe sans-cle pouvant etre issue d'une regle de creation de table,
+				// auquel cas les cles ne sont pas necessaire y compris dans le cas d'un flocon creer par des regles
```

Comment:
s/creer/cree/ (past participle instead of infinitive).

Reply:
Corrigé